308 automatic generation of python binding code (#333)
DLR-SC/memilio: pycode/memilio-generation/README.md

# MEmilio Automatic Code Generation of Python Bindings

This package contains Python bindings for the MEmilio C++ library. 
It enables the automatic generation of a part of the [Python Bindings](../memilio-simulation/README.md). It focuses on the model specific binding files e.g., for the seir model the files `oseir.cpp` and `oseir.py`.

The software was developed as a part of the bachelor thesis [Automatische Codegenerierung für nutzerfreundliche mathematisch-epidemiologische Modelle](https://elib.dlr.de/). For further insigths of the concept look into chapter 5.

## Prerequisites

The package uses the [Clang C++ library](https://clang.llvm.org/) and the [LibClang Python library](https://libclang.readthedocs.io/en/latest/index.html) to analyze the C++ code of a model. Both need to be installed and share the same version.

> To use the package you need a usable model including a compilation database (compile_commands.json). For models of the C++ library the compilation database is generated when building the [C++ Library](../../cpp/README.md).

Author: You are right. I have corrected this part in the README.

On the second part: Im not sure how that affects the building process.

The package provides an example script on how to use it in `memilio/tools`. The example uses the ode_seir model from the [C++ Library](../../cpp/models/ode_seir/README.md).

Before running the example you have to do these steps of setup:
- Change [config.json](./memilio/tools/config.json). Documentation in the ScannerConfig class.
- Change [ScannerConfig class](./memilio/generation/scanner_config.py).

Run the example with the command (path according to the current folder):

```bash
python memilio/tools/seir.py 
```

When working on a new model you can copy the example script and add an additional segment to the config.json. The setup works similar to the example. Additionaly you can print the AST of your model into a file (Usefull for development/debugging).

## Testing

The package provides a test suite in `memilio/generation_test`. To run the tests, simply run the following command

```bash
python -m unittest
```

## Development

When implementing new model features you can follow these steps:
- Add necessary configurations to [config file](./memilio/tools/config.json) and add corresponding attributes to the [ScannerConfig class](./memilio/generation/scanner_config.py).

> - Find the nodes in the AST with the features you want to implement (use method Scanner.output_ast_file()).

Reviewer: ```suggestion
- Find the nodes in the abstract syntax tree (AST) with the features you want to implement (use method Scanner.output_ast_file()).
```

Maybe 1-2 more sentence what these nodes could be?

# MEmilio Automatic Code Generation of Python Bindings

This package contains Python bindings generating code for the MEmilio C++ library. 
It enables the automatic generation of a part of the [Python Bindings](../memilio-simulation/README.md) that is common across multiple models. For a particular example, see the SEIR model with its files `oseir.cpp` and `oseir.py`.

> This generating software was developed as a part of the Bachelor thesis [Automatische Codegenerierung für nutzerfreundliche mathematisch-epidemiologische Modelle](https://elib.dlr.de/190367/). For further insigths of the concept look into Chapter 5.

Author: Issue #566

The package provides an example script on how to use it in `memilio/tools`. The example uses the ode_seir model from the [C++ Library](../../cpp/models/ode_seir/README.md).

Before running the example you have to do these steps of setup:
- Change [config.json.txt](./memilio/tools/config.json.txt). You can find a documentation in the ScannerConfig class.

> - Look into the parameters set in __post_init__() of the [ScannerConfig class](./memilio/generation/scanner_config.py), if they match up with the cpp-class names.

Reviewer: ```suggestion
- Check if the parameters set in __post_init__() of the [ScannerConfig class](./memilio/generation/scanner_config.py) match with the cpp-class names.
```
